beacon_node/lighthouse_network/src/types/topics.rs

```diff
@@ -24,70 +28,138 @@ pub const BLS_TO_EXECUTION_CHANGE_TOPIC: &str = "bls_to_execution_change";
+}
+
+/// Returns all the topics the node should subscribe at `epoch`
+pub fn core_topics_to_subscribe<E: EthSpec>(
+    spec: &ChainSpec,
+    epoch: Epoch,
+    opts: &TopicSubscriptionOpts,
+) -> Vec<GossipKind> {
+    let fork_name = spec.fork_name_at_epoch(epoch);
+
+    let mut topics = vec![
+        GossipKind::BeaconBlock,
+        GossipKind::BeaconAggregateAndProof,
+        GossipKind::VoluntaryExit,
+        GossipKind::ProposerSlashing,
+        GossipKind::AttesterSlashing,
+    ];
+
+    if opts.subscribe_all_subnets {
+        for index in 0..spec.attestation_subnet_count {
+            topics.push(GossipKind::Attestation(index.into()));
+        }
+    }
+
+    if fork_name >= ForkName::Altair {
+        topics.push(GossipKind::SignedContributionAndProof);
+
+        if opts.subscribe_all_subnets {
+            for index in 0..E::SyncCommitteeSubnetCount::to_u64() {
+                topics.push(GossipKind::SyncCommitteeMessage(index.into()));
             }
-            let mut deneb_topics = DENEB_CORE_TOPICS.to_vec();
-            deneb_topics.append(&mut deneb_blob_topics);
-            deneb_topics
         }
-        ForkName::Electra => vec![],
+
+        if opts.enable_light_client_server {
+            topics.push(GossipKind::LightClientFinalityUpdate);
+            topics.push(GossipKind::LightClientOptimisticUpdate);
+        }
+    }
+
+    if fork_name >= ForkName::Capella {
+        topics.push(GossipKind::BlsToExecutionChange);
     }
+
+    if fork_name >= ForkName::Deneb && !spec.is_peer_das_enabled_for_epoch(epoch) {
+        // All of deneb blob topics are core topics
+        for i in 0..spec.blob_sidecar_subnet_count {
+            topics.push(GossipKind::BlobSidecar(i));
+        }
+    }
+
+    if spec.is_peer_das_enabled_for_epoch(epoch) {
+        if opts.subscribe_all_data_column_subnets {
+            for index in 0..spec.data_column_sidecar_subnet_count {
+                topics.push(GossipKind::DataColumnSidecar(index.into()));
+            }
+        } else {
+            for &index in &opts.custody_subnets {
+                topics.push(GossipKind::DataColumnSidecar(index));
+            }
+        }
+    }
+
+    topics
 }
 
-/// Returns all the attestation and sync committee topics, for a given fork.
-pub fn attestation_sync_committee_topics<E: EthSpec>() -> impl Iterator<Item = GossipKind> {
-    (0..E::SubnetBitfieldLength::to_usize())
-        .map(|subnet_id| GossipKind::Attestation(SubnetId::new(subnet_id as u64)))
-        .chain(
-            (0..E::SyncCommitteeSubnetCount::to_usize()).map(|sync_committee_id| {
-                GossipKind::SyncCommitteeMessage(SyncSubnetId::new(sync_committee_id as u64))
-            }),
-        )
+pub fn new_topics_to_subscribe_at_epoch<E: EthSpec>(
+    spec: &ChainSpec,
+    next_epoch: Epoch,
+    opts: &TopicSubscriptionOpts,
+) -> Vec<GossipKind> {
+    let prev_epoch = next_epoch.saturating_sub(Epoch::new(1));
```

Comment:
I'm not sure if this method is called at genesis, but if it is, I don't think it works since you'll never have a different `prev_set` and `next_set`

Reply:
Nice pick! Fixed in https://github.com/sigp/lighthouse/pull/5899/commits/6fa81e819d46542a6ac047eef365c39246fffca9